### PR TITLE
WIP: Data Layer - Generic API data system

### PR DESCRIPTION
**Pull request by dmsnell**
**Quick highlights**
 - Add data from an API endpoint in 5-10 lines of code
 - Reduce size of data layer and network code by _ripping it out_
 - Freshness, request status, "static typing" data models
 - Declarative data needs without query components

**Current state of this PR**
This PR is in the _doodle_ state where no code is expected to be
run and it's more of a shared scratchpad for discussing the ideas.
Expect it to be full of holes.

**Questions**
 - Dynamic state paths are tricky because the knowledge we have to _request_ data may be different than what we get when _receiving_ data.
 - State retrieval is easy because we can `get()` it, but `set()` mutates. We need to build an immutable updater that can take a dynamic or stringy path and return a new object with that deep property updated. We may consider some kind of DSL such as React's `immutability-helper` uses or consider requiring a function of some kind.
 - This kind of prevents splitting separate data types when they come from one endpoint. However, the point of "shoving" is that we have already been neglecting these kinds of relationships in our state code.
 - As usual, I think much more about _getting_ data than _updating_ it. We should be able to plainly update data here by switching to a `POST` and submitting the resource in question, but I could use some input on how to make this work well.

**Running Idea**

Imagine you are displaying a post. You can require the data through `connect()` without changing anything but the version of `connect()`:

```js
import { connect } from 'state/api-data';

const Post = ( 
	author: { displayName },
	isLoading,
	post: { title, content },
	siteId,
	postId,
) => (
	<div>
		<h1>{ title }</h1>
		<h2>{ displayName }</h2>
		{ isLoading( 'POST', { siteId, postId } )
			? <Placeholder />
			: <div>{ content }</div>
		}
	</div>
);

export default connect( ( state, { userId, siteId, postId } ) => ( {
	apiData: {
		author: [ 'WP_USER', userId ],
		post: [ 'POST', { siteId, postId }, {
			fresherThan: MINUTE_IN_SECONDS, // overwrite default options
			forceUpdates: true, // keep polling even when we haven't re-rendered
		} ],
	},
	siteName: getSelectedSite( state, siteId ).title,
} ) )( Post );
```

In this component Calypso will automatically make sure that the requested information is requested from the network and kept up-to-date based on some expiry/freshness data specified in the `api-data` data type definitions. What does that look like?

```js
const postTypes = {
	POST: {
		// creates an HTTP/WP_HTTP action to fetch the data
		// no need to provide `onSuccess` and friends
		fetch: ( { siteId, postId } ) => http( {
			method: 'GET',
			apiVersion: '1.2',
			path: `/sites/${ siteId }/posts/${ postId }`
		} ),

		// where in the apiData state tree this stuff lives
		// lodash like?
		dataPath: ( { siteId, postId } ) => `${ siteId }.${ postId }`,
	},
	POSTS: {
		fetch: siteId => http( {
			method: 'GET',
			apiVersion: '1.2',
			path: `/sites/${ siteId }/posts`
		} ),
		dataPath: siteId => [ siteId ],
		whenMissing: DO_NOT_REMOVE,
	},
	…
};
```

Of course, we have many different options available to us for additional stuff like freshness and what to do when we get a collection of things and what to do when items from missing from newer API responses that were already in state etc… all of these are details I think will be easy to add as we better understand our needs.

> This system heavily relies on the foundation which the data layer has established for decoupling network calls from their fulfillment and for separating requests of data with the actual fetching of that data.

---

A while back in a discussion with @gziolo I tossed around an idea for
dynamically creating the code needed to "shove" data from an API request
into Redux state. The purpose of the conversation was to discuss the
growing bulk of code doing nothing more than "shoving." While I think
that this practice of "shoving" is missing out on many advantages that
Redux has to offer, the more I dove into this hypothetical system for
interacting with "API data" the more I realized we could have real
pragmatic benefits from building it out:

 - Significantly less code for adding data from new API endpoints
 - The ability to declaratively list data needs at the UI level without
   query components
 - A common and standardized way of interacting with remote data
 - Strongly-typed data models

So in time I drafted an API spec for how I would like to interact with
such a system and doodled the following snippet:

```js
const apiTypes = {
  [ GEO ]: [ apiDump(
    // creates the request object
    () => rawHttp( { path: 'public-api.wordpress.com/geo' } ),

    // base path for sub-tree (top-level state)
    '',

    // array of pairs of sub-paths and values to inject
    // will end up at 'state.geo'
    ( { geo } ) => [ [ 'geo', geo ] ],

    // optional parameters
    {
      fresherThan: DAY_IN_SECONDS,
      onFailure: () => errorNotice( translate(
        'Could not fetch geolocation information, sorry.'
      ) ),
    }
  ) ],

  [ ZORPS ]: [ apiDump(
    // creates the request object
    ( { siteId } ) => http( { path: `​/sites/${ siteId }/zorps`, apiVersion: 'v1', method: 'GET' } ),

    // base path for sub-tree
    'misc.zorps',

    // array of pairs of sub-paths and values to inject
    // will end up at 'state.misc.zorps[ zorp.id ]'
    ( { zorps } ) => zorps.map( zorp => [ zorp.id, zorp ] ),
  ) ],
}

// use it
const geo = apiData( GEO );
const zorp = apiData( ZORPS, { id: zorpId } );

const isRequestingGeo = isApiRequesting( GEO );
const hasValidZorps = didApiFail( ZORPS );

dispatch( requestApiData( GEO ) );
dispatch( requestApiData( ZORPS, { siteId } ) );
```

This "doodle" demonstrates one way we could tell Calypso about the new
types of data we're connecting (`apiTypes`), how to access that data
from a UI component (`apiData()`), how to determine meta informationa
about the data (`isApiRequesting()`, `didApiFail()`), and how to request
the data (`requestApiData()`).

Since the time I wrote the doodle the ideas have been stewing in my head
and I decided that it should probably be "mandatory" to provide a type
of explicit, whitelisted, schema for the API responses so that we can
have an official definition of the type and shape of the data we're
sucking into the application.

As I started writing up this PR (frankly another code doodling session)
I made some further refinements and discovered some additional questions
that we'll need to deal with in order to make this happen.

Please see the README in the PR for the ongoing explanation and
motivation for this new data system.

### CODE REVIEW

**Comment by aduth** [comment]
A few thoughts that come up as I look through this:

In your example you have `isLoading` coming from a named import, and in the code as a prop. I would expect with the named import we might run into challenges about how exactly that status is determined, since it would assume some global state of requests. If this is meant to support updating in addition to fetching, we should also consider whether these statuses are reflected in distinct prop names, or arguments to the status getter.

With current query components, we are able to tie data fetching to lifecycle of a component. A freshness threshold can help eliminate some redundant requests, but I think we'd still want to be able to support (default to?) forcing the fetch on mount. The thinking being that as the user navigates around the application, they should always see fresh data. For example, if I trash a post, then switch to the "Trash" filter, I shouldn't not see the post I just trashed just because the last fetch was fresh enough. Of course, in this context we could know how to transform state in response to the user trashing action ([which is a problem Query Manager was meant to solve](https://github.com/Automattic/wp-calypso/tree/master/client/lib/query-manager#query-manager)). Updates from other users is trickier, though perhaps similarly solvable (live updates via websockets).

Trying to understand the `dataPath`. For something like `POSTS` defining a `dataPath` of `dataPath: siteId => [ siteId ],`, how is this deduplicated with, say, a site endpoint? Would we generate a key to group this under based on the name of the data type? At that point, could we consider generating the entire data path automatically based on the incoming "id" and avoid an explicit `dataPath` altogether?

Also going to point to [Gutenberg's `withAPIData`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data), which makes many compromises so far as separating data shape from network requests, but shares a few of the same characteristics of connecting individual components. One nice thing about your approach for enhancing the default react-redux `connect` is that you can pass as arguments to the API connector values from state without much trouble. As a separate higher-order component, this is still reasonable, though requiring [extra work](https://github.com/WordPress/gutenberg/blob/67e759b332e284ca541d82e9af0b3367a92d2347/editor/sidebar/last-revision/index.js#L49-L66) ([perhaps also counter-intuitive](https://github.com/WordPress/gutenberg/pull/2500#discussion_r134735214)).

**Comment by dmsnell** [comment]
> I would expect with the named import we might run into challenges about how exactly that status is determined

this has been one of the more challenging bits for me @aduth - I don't think we can pull it in as a named import. there are good questions about how to include it in `props` and that's related to other performance considerations I think we need to be aware of (how this affects shallow comparisons for re-rendering) but I think they are all workable. I have updated the example

**Comment by dmsnell** [comment]
> With current query components, we are able to tie data fetching to lifecycle of a component. A freshness threshold can help eliminate some redundant requests, but I think we'd still want to be able to support (default to?) forcing the fetch on mount.

This is another good question. I feel like the bit about updating the posts probably has more to do with trashing the post than viewing a list of posts. If we force updates on mount then we're doing that whether or not a post was trashed and that seems like a wasteful default.

If we're really using Redux properly then we'll be updating posts on the related actions, like trash, and it won't be a problem. As we "shove API data" more we lose those related connections and this becomes more of an issue. On the other hand, we should be able to use the functionality of the data layer and "freshness" to handle this more elegantly in an automated fashion. Consider this hypothetical data layer handler…

```js
[ POST_TRASH ]: [ ( store, { siteId } ) => withFreshness( 0, { type: POSTS_REQUEST, siteId } ) ) ]
```

Whenever we see a post get trashed we dispatch a single request to update post lists with freshness of `0`, which forces an immediate update. I _think_ that addresses what you are asking about _plus_ it leaves the relations in a somewhat centralized or descriptive spot.

**Comment by samouri** [comment]
I love this idea. Radically reducing the amount of code we write while also tightening up  correctness and expressiveness is a massive win all around.

I'm hoping we can build up to something like this incrementally instead of all-at-once.  The grand vision of ~5-10 LOC for generating the data-layer, selectors, and reducers in a flexible way is our north star, and these code scratchpads are helpful for getting us all on the same page.  Here are the primitives I'm imagining:
1. generic higher order reducers (HoRs) that we can start replacing current reducers with
2. autogenerating selectors based on the the HoR and state-path
3. incorporating freshness  into the current data-layer and HoRs as properties of the request action/result
4.  any others i'm missing?

after we have those pieces in place I think we can feel confident that we've built the right abstractions necessary for the generic api data system.

**Comment by dmsnell** [comment]
> Trying to understand the dataPath. For something like POSTS defining a dataPath of dataPath: siteId => [ siteId ], how is this deduplicated with, say, a site endpoint? Would we generate a key to group this under based on the name of the data type? At that point, could we consider generating the entire data path automatically based on the incoming "id" and avoid an explicit dataPath altogether?

This is a point I'm having trouble understanding. We definitely could auto-generate keys based on something like `deterministicStringify()` but one thing I thought might be more valuable is being able to spread out collections of things in API responses. For example, consider getting a list of posts…

```js
dataPath: ( { found, posts } ) => [
	[ 'totalPosts', found ],
	...posts.map( post => [ post.id, post ] )
]
```

The idea here was we can return a list of pairs of paths and data that would fill a collection in state and it would be based entirely off of the `id` and the data itself. This has been complicated to do though.

I want it all to be expressible from the `id` but have response data map appropriately…

**Comment by dmsnell** [comment]
> after we have those pieces in place I think we can feel confident that we've built the right abstractions necessary for the generic api data system.

I'm questioning this as an incremental thing or even as a normative thing. We lose a lot of value when we isolate API endpoints as data types.

Primarily this is a response from several instances of this code already existing in Calypso. You can spot them because the reducers all look like this:

```js
export const reducer = ( state = {}, action ) => {
	if ( action.type === THING_RECEIVE ) {
		return {
			...state,
			[ action.siteId ]: action.payload,
		}
	}

	return state;
}
```

So these things _already_ have a ton of code which won't actually benefit much from data layer API updates, because it's already doing the repetitive and minimal bits. For these cases I'd rather just identify the pattern they are all using and rip it out as a complementary system to more general interactions in the data layer.

In other words, I don't see any incremental paths to this because it's going in a direction the data layer isn't or maybe shouldn't.

> after

why do you think this should come _after_?

**Comment by samouri** [comment]
> why do you think this should come after?

because I think the generic data system could be constructed from the primitives I outlined.

> I don't see any incremental paths to this because it's going in a direction the data layer isn't or maybe shouldn't.


I don't pretend to have thought up working api contracts,
but here is my general train of thought for how we could do something like this iteratively:

**1. generic HoRs**

```js
/* 
 * a slightly configurable reducer for handling our basic reducers
 */
function genericReducer( { listenFor, keyGenerator, mergeBy, } ) {
  ...
  const reducer = (state, action) =>  action.type === listenFor
    ? mergeBy( { state, keyGenerator } )
    : state;
  return reducer;
}

// these are potential values for the mergeBy argument of genericReducer
const MergeTypes = {
  replaceEntirely() { ... } 
  shallowMergeById() { ... }
  deepMergeById() { ... }
}
```

If the handwavy code above was ironed out, then we could replace the THING_RECEIVE reducer with:

```js
export const reducer = genericReducer( { 
  listenFor: 'THING_RECEIVE', 
  mergeBy: MergeTypes.shallowMergeById,
  keyGenerator: action => action.siteId
} )
```

This change would benefit both current writers of reducers and also can eventually be used by the 
generic data system. 

**2. auto-generate selectors**

to be able to automatically generate a selector we will need to know two things:
1. the path within state (statePath) where the reducer lives
2. how to key within that reducer (I've been calling it the keyGenerator)

One spitball idea for how we could iteratively create this would be to incorporate selector generation into `genericReducer` when a statePath is provided:  
```js
function genericReducer( { listenFor, keyGenerator, mergeBy, statePath } ) {
  ...
  if ( statePath ) {
    reducer.selector = ( state, params ) => 
       get( state, statePath )[ keyGenerator( params ) ]
  }
  ...
}
```

Now that the selector is automatically generated and we are flexible with how to export/import it.

**3. adding in freshness**

An important note here is that I think data-freshness should be added to the current implementation of the data-layer. 
It can be configured within the actions's meta key.  This way we can then add the functionality into existing actions and query-components.

We'd also need to create a HoR for augmenting current reducers with freshness data that lives in action meta.  We can then bake it back into genericReducer (i know i know, we ideally could keep freshness data outside of redux)

```js
/*
 * HoR for augmenting items with api freshness data
 * 
 * precondition:
 *    this assumes the reducer is keyed and that each leaf item is itself 
 *    an object that can be augmented with freshness data obtained from the action
 *    meta.  i.e. { type: 'XYZ', meta: { dataLayer: { freshness: { timeReceived: 42 } } } }
 */ 
function withFreshness( { reducer,  keyGenerator }) {
 ...
}

function genericReducer( ... ) {
 ...
  return withFreshness( { reducer, keyGenerator } );
}
```

Note: i have a feeling we'd actually want to keep freshness data within the data-layer, but right now i'm not sure 
how we could reasonably accomplish that while allowing ui components to both force-request onMount and freshness poll when rendered.


**4. combine it all for generic api data system**

to be continued

**Comment by apeatling** [comment]
@dmsnell It's been some time here, should we close this out and have the calypso team help out with dev and reviews once we get further into projects?

**Comment by dmsnell** [comment]
@apeatling a discussion would be helpful. it's been a long time but there hasn't been too much review here. I'd like to keep it open while working on #22549 and the new `npm` package I'm working on from these ideas. I have referred to them multiple times and try to keep them rebased on recent **master** commits

**Comment by dmsnell** [comment]
Closing for a future iteration based on #22549 - thanks all who provided feedback here on the design.
